pkg/apis/admission/types.go

```diff
@@ -0,0 +1,75 @@
+limitations under the License.
+*/
+
+package admission
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/kubernetes/pkg/apis/authentication"
+)
+
+// AdmissionReview describes an admission request.
+type AdmissionReview struct {
+	metav1.TypeMeta
+
+	// Spec describes the attributes for the admission request.
+	// Since this admission controller is non-mutating the webhook should avoid setting this in its response to avoid the
+	// cost of deserializing it.
+	Spec AdmissionReviewSpec
+	// Status is filled in by the webhook and indicates whether the admission request should be permitted.
+	Status AdmissionReviewStatus
+}
+
+// AdmissionReviewSpec describes the admission.Attributes for the admission request.
+type AdmissionReviewSpec struct {
+	// Kind is the type of object being manipulated.  For example: Pod
+	Kind metav1.GroupVersionKind
+	// Name is the name of the object as presented in the request.  On a CREATE operation, the client may omit name and
+	// rely on the server to generate the name.  If that is the case, this method will return the empty string.
+	Name string
+	// Namespace is the namespace associated with the request (if any).
+	Namespace string
+	// Object is the object from the incoming request prior to default values being applied
+	Object runtime.Object
+	// OldObject is the existing object. Only populated for UPDATE requests.
+	OldObject runtime.Object
+	// Operation is the operation being performed
+	Operation admission.Operation
+	// Resource is the name of the resource being requested.  This is not the kind.  For example: pods
+	Resource metav1.GroupVersionResource
+	// SubResource is the name of the subresource being requested.  This is a different resource, scoped to the parent
+	// resource, but it may have a different kind. For instance, /pods has the resource "pods" and the kind "Pod", while
+	// /pods/foo/status has the resource "pods", the sub resource "status", and the kind "Pod" (because status operates on
+	// pods). The binding resource for a pod though may be /pods/foo/binding, which has resource "pods", subresource
+	// "binding", and kind "Binding".
+	SubResource string
+	// UserInfo is information about the requesting user
+	UserInfo authentication.UserInfo
+}
+
+// AdmissionReviewStatus describes the status of the admission request.
+type AdmissionReviewStatus struct {
+	// Allowed indicates whether or not the admission request was permitted.
+	Allowed bool
+	// A human-readable description of the status of this operation.  Kubernetes may truncate excessively long errors when
+	// displaying to the user.
+	Message string
+	// A machine-readable description of why this operation is in the state it is. If this value is empty there is no
+	// information available.  A Reason clarifies an HTTP status code but does not override it.
+	Reason metav1.StatusReason
```

Comment:
Admission is basically `func(Attributes) error`.  If `spec` is attributes, then this (status) needs to be a `metav1.Status`, probably modeled as:

```
type AdmissionReviewStatus  struct {
  Allowed bool
  // Reason is set if the webhook wishes to return additional information about the result of this operation. If Allowed is true, this field is informational only. If Allowed is false, this can alter the error an end user sees.
  Reason *metav1.Status  
}
```

I'm torn on `Reason` as the field name - I'm somewhat concerned with stuttering because Status already contains fields that match all possible names, i.e.:

* status.status.reason (redundant)
* status.reason.reason (redundant)
* status.details.details (redundant)
* status.error.details (we don't use "error" anywhere in the API today, and I'm hesitant to make the status field, which can represent success as well, also be negative)